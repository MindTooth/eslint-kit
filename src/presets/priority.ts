import { PresetName } from './presets'

export const PRIORITY: PresetName[] = [
  // should be applied first
  'typescript',

  'prettier',
  'node',
  'react',
  'solid.js',

  // apply after react preset
  'next.js',

  // may need next.js integration in future
  'effector',

  // apply after all extensions are set
  'alias',
]
