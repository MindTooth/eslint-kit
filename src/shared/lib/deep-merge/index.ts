/* eslint-disable @typescript-eslint/no-explicit-any */
import { DeepMerge, PathsOf } from './types'

function isObject(value: any): value is Record<string, any> {
  return typeof value === 'object' && value !== null
}

export enum Merge {
  Shallow,
  Deep,
  Override,
}

type AllPaths<A, B> = PathsOf<A> | PathsOf<B> | null

// eslint-disable-next-line sonarjs/cognitive-complexity
export function deepMerge<A extends any, B extends any>(
  a: A,
  b: B,
  strategy: (path: AllPaths<A, B>) => Merge = () => Merge.Deep,
  basePath: string | null = null,
  root = true,
  shallow = false
): DeepMerge<A, B> {
  if (Array.isArray(a) && Array.isArray(b)) {
    return [...a, ...b] as DeepMerge<A, B>
  }

  if (isObject(a) && isObject(b)) {
    if (shallow) {
      return { ...a, ...b } as DeepMerge<A, B>
    }

    if (root) {
      if (strategy(null) === Merge.Shallow) {
        return deepMerge(a, b, strategy as any, null, false, true) as DeepMerge<
          A,
          B
        >
      }

      if (strategy(null) === Merge.Override) {
        return b as DeepMerge<A, B>
      }
    }

    const keys = [...Object.keys(a), ...Object.keys(b)]
    const final: any = {}

    for (const key of keys) {
      const path = basePath ? [basePath, key].join('.') : key

      if (strategy(path as any) === Merge.Shallow) {
        final[key] = deepMerge(
          a[key],
          b[key],
          strategy as any,
          path,
          false,
          true
        )
      } else if (strategy(path as any) === Merge.Override) {
        final[key] = b[key]
      } else {
        final[key] = deepMerge(a[key], b[key], strategy as any, path, false)
      }
    }

    return final
  }

  if (typeof b !== 'undefined') {
    return b as DeepMerge<A, B>
  }

  return a as DeepMerge<A, B>
}
