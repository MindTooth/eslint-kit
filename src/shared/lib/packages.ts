import { parse } from 'semver'

type Dependencies = Record<string, string>

export interface PackageJson {
  version: string
  dependencies?: Dependencies
  devDependencies?: Dependencies
}

function mergeDependencies(packageJson: PackageJson) {
  const { dependencies = {}, devDependencies = {} } = packageJson
  return { ...dependencies, ...devDependencies }
}

export function getPackageSemver(name: string, packageJson: PackageJson) {
  const allDependencies = mergeDependencies(packageJson)
  const version = allDependencies[name]
  if (!version) return null
  const parsed = parse(version)
  if (!parsed) return null
  return parsed
}
