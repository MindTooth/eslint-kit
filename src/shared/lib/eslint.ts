import { Linter } from 'eslint'
import { deepMerge, Strategy } from './deep-merge'

function parseExtension(pattern: string): string | null {
  if (!pattern.startsWith('*.')) return null
  return pattern.slice(2)
}

function splitOverrides(
  overrides: Linter.ConfigOverride[]
): Linter.ConfigOverride[] {
  const result: Linter.ConfigOverride[] = []

  for (const { files, ...rest } of overrides) {
    const patterns = Array.isArray(files) ? files : [files]
    for (const pattern of patterns) {
      result.push({ files: [pattern], ...rest })
    }
  }

  return result
}

export function overrideOverrides(
  overrides: Linter.ConfigOverride[]
): Linter.ConfigOverride[] {
  const result: Linter.ConfigOverride[] = []

  const reversedOverrides = overrides.slice().reverse()

  const starRules = new Set<string>()
  const usedRules = new Map<string, Set<string>>()

  for (const { files, rules = {}, ...rest } of reversedOverrides) {
    const [pattern] = Array.isArray(files) ? files : [files]

    const extension = parseExtension(pattern)
    if (!extension) continue

    if (!usedRules.has(extension)) {
      usedRules.set(extension, new Set<string>())
    }

    const extensionUsedRules = usedRules.get(extension) ?? new Set()
    if (!usedRules.has(extension)) usedRules.set(extension, extensionUsedRules)

    const overridedRules: Linter.RulesRecord = {}

    for (const [ruleName, ruleEntry] of Object.entries(rules)) {
      if (!ruleEntry) continue
      if (extensionUsedRules.has(ruleName)) continue
      if (starRules.has(ruleName)) continue
      overridedRules[ruleName] = ruleEntry
      extensionUsedRules.add(ruleName)
      if (extension === '*') starRules.add(ruleName)
    }

    result.push({ files, rules: overridedRules, ...rest })
  }

  return result
}

export function mergeConfigs(configs: Linter.Config[]): Linter.Config {
  const config = configs.reduce((final, current) => {
    return deepMerge(final, current, (path) => {
      if (path === 'rules') return Strategy.Shallow
      return Strategy.Deep
    })
  })

  const splittedOverrides = splitOverrides(config.overrides ?? [])
  const overrides = overrideOverrides(splittedOverrides)

  return { ...config, overrides }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type OnlyCollections<T> = T extends any[] ? T : T extends object ? T : never

function applier<
  T extends 'rules' | 'settings' | 'extends' | 'parserOptions' | 'overrides'
>() {
  return (
    condition: unknown,
    entity: OnlyCollections<Linter.Config[T]>
  ): OnlyCollections<Linter.Config[T]> => {
    const fallback = Array.isArray(entity) ? [] : {}
    if (!condition) return fallback as OnlyCollections<Linter.Config[T]>
    return entity
  }
}

export const conditional = {
  rules: applier<'rules'>(),
  settings: applier<'settings'>(),
  extends: applier<'extends'>(),
  parserOptions: applier<'parserOptions'>(),
  overrides: applier<'overrides'>(),
}
