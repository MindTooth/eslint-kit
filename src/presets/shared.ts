import { Linter } from 'eslint'
import { mergeConfigs } from '../shared/lib/eslint'
import { Jsconfig, PackageJson } from '../shared/types'
import { PresetName } from './names'

type ReplaceVoid<T> = void extends T ? Exclude<T, void> | undefined : T

export interface Meta {
  root: string
  readPackageJson(): PackageJson | null
  readJsconfig(): Jsconfig | null
  presets: Set<PresetName>
  imports: {
    extensions: string[]
  }
  typescript: {
    root: string
    tsconfig: string
  }
}

export interface Input<T = void> {
  options: ReplaceVoid<T>
  meta: Meta
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export interface Preset<T = any> {
  name: PresetName
  options: T
  updateMeta(input: Input<T>): void
  compile(input: Input<T>): Linter.Config
}

export type PresetFabric<T = void> = (options: T) => Preset<T>

interface CreatePresetParams<T = void> {
  name: PresetName
  updateMeta?(input: Input<T>): void
  compile(input: Input<T>): Linter.Config
}

export function createPreset<T = void>({
  name,
  updateMeta = () => {},
  compile,
}: CreatePresetParams<T>): PresetFabric<T> {
  return (options) => ({
    name,
    options,
    updateMeta,
    compile,
  })
}

export const createMeta = (): Meta => ({
  root: process.cwd(),
  readPackageJson: () => null,
  readJsconfig: () => null,
  presets: new Set(),
  imports: {
    extensions: [],
  },
  typescript: {
    root: './',
    tsconfig: 'tsconfig.json',
  },
})

export function compilePresets(
  presets: Preset[],
  priority: string[]
): Linter.Config {
  const prioritized = presets
    .slice()
    .sort((a, b) => priority.indexOf(a.name) - priority.indexOf(b.name))

  const meta = createMeta()

  for (const preset of prioritized) {
    meta.presets.add(preset.name)
    preset.updateMeta({ meta, options: preset.options })
  }

  const configs = prioritized.map((preset) =>
    preset.compile({ meta, options: preset.options })
  )

  return mergeConfigs(configs)
}
