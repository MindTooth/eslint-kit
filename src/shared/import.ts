import { Linter } from 'eslint'
import { extensions } from './constants'

export const importSettings: Linter.Config['settings'] = {
  'import/extensions': extensions.js,
  'import/resolver': {
    node: {
      extensions: extensions.js,
    },
  },
  'import/core-modules': [],
  'import/ignore': ['\\.(coffee|scss|css|less|hbs|svg|json)$'],
  'import/internal-regex': '^@types/',
}

type ImportExtensionsRule = [
  Linter.RuleLevel,
  'ignorePackages',
  Record<string, 'never'>
]

export const importExtensionsRule: ImportExtensionsRule = [
  'warn',
  'ignorePackages',
  {
    js: 'never',
    mjs: 'never',
    jsx: 'never',
  },
]

export const importRules: Linter.RulesRecord = {
  'import/no-unresolved': ['error', { commonjs: true, caseSensitive: true }],
  'import/no-anonymous-default-export': [
    'error',
    {
      allowCallExpression: false,
    },
  ],
  'import/export': 'error',
  'import/no-mutable-exports': 'error',
  'import/no-amd': 'error',
  'import/first': 'error',
  'import/no-duplicates': 'error',
  'import/extensions': importExtensionsRule,
  'import/newline-after-import': 'warn',
  'import/no-webpack-loader-syntax': 'error',
  'import/no-self-import': 'error',
  'import/no-cycle': ['error', { maxDepth: Number.POSITIVE_INFINITY }],
  'import/no-useless-path-segments': ['warn', { commonjs: true }],
}
