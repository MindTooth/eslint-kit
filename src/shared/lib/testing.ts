import fs from 'fs'
import path from 'path'
import { ESLint, Linter } from 'eslint'

export interface Options {
  config: unknown
  basePath: string
  files: string[]
  extension: 'js' | 'ts' | 'tsx'
}

export async function testConfig({
  config,
  basePath,
  files,
  extension,
}: Options) {
  const cli = new ESLint({
    baseConfig: config as Linter.Config<Linter.RulesRecord>,
    useEslintrc: false,
    cwd: process.cwd(),
    ignore: false,
  })

  for (const file of files) {
    const filePath = path.resolve(basePath, `./tests/${file}.${extension}`)
    const code = fs.readFileSync(filePath).toString()
    const isTS = ['ts', 'tsx'].includes(extension)
    const result = await cli.lintText(code, {
      filePath: isTS ? filePath : undefined,
    })
    expect(result).toMatchSnapshot(`${file}`)
  }
}
