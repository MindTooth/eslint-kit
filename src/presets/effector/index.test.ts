import { testConfig } from '../../shared/lib/testing'
import { base } from '../base'
import { PRIORITY } from '../priority'
import { compilePresets } from '../shared'
import { effector } from './index'

describe('[Presets] React', () => {
  it('should match snapshots', async () => {
    await testConfig({
      config: compilePresets([base(), effector()], PRIORITY),
      basePath: __dirname,
      files: ['basic-clear', 'basic-warn', 'basic-error'],
      extension: 'js',
    })
  })
})
