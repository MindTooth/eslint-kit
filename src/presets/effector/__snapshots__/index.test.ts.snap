// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Presets] Effector should match snapshots: basic-clear 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/effector/tests/basic-clear.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [],
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`[Presets] Effector should match snapshots: basic-error 1`] = `
Array [
  Object {
    "errorCount": 1,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/effector/tests/basic-error.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [
      Object {
        "column": 23,
        "endColumn": 54,
        "endLine": 6,
        "line": 5,
        "message": "Method \`.on\` is called on store \`$users\` more than once for userAdded.",
        "messageId": "duplicateOn",
        "nodeType": "CallExpression",
        "ruleId": "effector/no-duplicate-on",
        "severity": 2,
      },
    ],
    "source": "import { createEvent, createStore } from 'effector'

const userAdded = createEvent()

export const $users = createStore([])
  .on(userAdded, (users, user) => users.concat(user))
  .on(userAdded, (users, user) => [...users, user])
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;

exports[`[Presets] Effector should match snapshots: basic-warn 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/effector/tests/basic-warn.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [
      Object {
        "column": 1,
        "endColumn": 3,
        "endLine": 8,
        "line": 3,
        "message": "Order of options should be clock->source->filter->fn->target, but found fn->target->clock->source",
        "messageId": "invalidOrder",
        "nodeType": "CallExpression",
        "ruleId": "effector/keep-options-order",
        "severity": 1,
        "suggestions": Array [
          Object {
            "desc": "Sort options",
            "fix": Object {
              "range": Array [
                69,
                168,
              ],
              "text": "{ clock: createEvent(), source: createStore(5), fn: value => value, target: createEvent() }",
            },
            "messageId": "changeOrder",
          },
        ],
      },
    ],
    "source": "import { createEvent, createStore, sample } from \\"effector\\";

sample({
  fn: value => value,
  target: createEvent(),
  clock: createEvent(),
  source: createStore(5)
})
",
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 1,
  },
]
`;
